### PR TITLE
Fix #7147

### LINKED ISSUE #7147: Expose total number of label pairs in head
## Proposal
**Use case. Why is this important?**

For scale analysis it would be great if we either exposed a metric that reflects the total number of labels (name + value combinations) in the head block. Potentially this could also be an entry in the head block cardinality section on the status page. I don't care how I just want to get the numbers out somehow and can't today.

Side note, yes top 10 cardinality labels are useful, but may be hiding a long tail of labels that still affect the head's index, but we can't be sure until we are able to validate with numbers.

Overall context: I'm researching some other indexing strategies and it's crucial to have some real world numbers for that.

@brian-brazil @roidelapluie @bwplotka @csmarchbanks @simonpasquier

### PR DESCRIPTION
**Pull request by nevill**
Add a new metric prometheus_tsdb_head_label_pairs, to count the unique number lable pairs in the head block. This will close #7147.

### CODE REVIEW

**Comment by nevill** [comment]
@brian-brazil can you review it?

**Comment by brian-brazil** [comment]
@codesome is the tsdb maintainer.

**Comment by nevill** [comment]
Sorry @brian-brazil :)
Hi @codesome can you take a review?

File: tsdb/head.go [hunk 1]
```diff
@@ -1327,7 +1334,13 @@ func (h *Head) gc() {
...
 			values[t.Name] = ss
 		}
-		ss.set(t.Value)
+
+		if _, ok := ss[t.Value]; !ok {
+			if t.Name != "" {
+				numLabelPairs++
+			}
+			ss.set(t.Value)
+		}
```
Comment (Harkishen-Singh) [inline tsdb/head.go:1343]: Not sure, but I think `getOrCreateWithID()` should be sufficient for this.

**Review by codesome — COMMENTED** [review]
Apologies for the super delayed review. Would you be able to make some changes?

Currently the implementation has some race. It would be much simpler if we keep the track of number of label pairs in `MemPostings` struct hence update it in `postings.Delete()` and `postings.Add()` call instead of tracking it in Head.

File: tsdb/head.go (hunk 1 above)
Comment (codesome) [reply tsdb/head.go:1343]: Only `getOrCreateWithID()` is not enough since it only accounts for addition of new label pairs and not deletion.

**Comment by nevill** [comment]
@codesome OK, glad to have some feedback, but I really need to refresh my mind to follow some recent changes of the project first.

**Comment by ankitpatel96** [comment]
bump - this would be a reallly really nice feature for us!

**Pull request closed without merging**